File: modules/storage.py
# :project: transfery
# :author: L-ING
# :copyright: (C) 2022 L-ING <hlf01@icloud.com>
# :license: MIT, see LICENSE for more details.

from miniopy_async import Minio
from miniopy_async.error import S3Error, InvalidResponseError
from miniopy_async.helpers import genheaders
from miniopy_async.datatypes import Part


class Storage:
    # Create a client with the MinIO server playground, its access key
    # and secret key.
    def __init__(self, host, username, password, bucket, secure):
        self.client = Minio(
            endpoint=host,
            access_key=username,
            secret_key=password,
            secure=secure
        )
        # for printing
        self.host = host
        self.bucket = bucket

    def init(self):
        try:
            if not self.client.bucket_exists(self.bucket):
                self.client.make_bucket(self.bucket)
        except InvalidResponseError as err:
            print(err)

    async def create_multipart_upload_id(self, remote_path):
        try:
            headers = genheaders(
                headers=None,
                sse=None,
                tags=None,
                retention=None,
                legal_hold=False
            )
            headers["Content-Type"] = "application/octet-stream"
            upload_id = await self.client._create_multipart_upload(self.bucket, remote_path, headers)
            return upload_id
        except S3Error as exc:
            print("error occurred.", exc)

    async def multipart_upload(self, remote_path, upload_id, part_data, part_number):
        try:
            etag = await self.client._upload_part(
                bucket_name=self.bucket,
                object_name=remote_path,
                data=part_data,
                headers=None,
                upload_id=upload_id,
                part_number=part_number
            )
            return etag
        except S3Error as exc:
            print("error occurred.", exc)

    async def complete_multipart_upload(self, remote_path, upload_id, parts):
        try:
            parts = [Part(part['partNumber'], part['etag']) for part in parts]
            await self.client._complete_multipart_upload(
                bucket_name=self.bucket,
                object_name=remote_path,
                upload_id=upload_id,
                parts=parts
            )
            print(
                "file is successfully uploaded as object %s to bucket %s." % (
                    remote_path,
                    self.bucket
                )
            )
            address = 'http://'+self.host+'/'+self.bucket+'/'+remote_path
            print(address)
        except S3Error as exc:
            print("error occurred.", exc)

    async def remove(self, remote_path):
        try:
            await self.client.remove_object(self.bucket, remote_path)
            print(
                "%s is successfully removed from bucket %s" % (
                    remote_path,
                    self.bucket
                )
            )
        except S3Error as exc:
            print("error occurred.", exc)

    async def remove_all(self):
        try:
            items = await self.list()
            for item in items:
                await self.remove(item['name'])
            print("all objects are removed from bucket %s successfully" % self.bucket)
        except S3Error as exc:
            print("error occurred.", exc)

    async def list(self):
        try:
            obj_list = await self.client.list_objects(self.bucket, recursive=True)
            obj_list = [
                {
                    'name': obj.object_name,
                    'size': obj.size,
                    'last_modified': obj.last_modified
                }
                for obj in obj_list
            ]
            print('objects list:', obj_list)
            return obj_list
        except S3Error as exc:
            print("error occurred.", exc)

    async def get_download_url(self, remote_path, change_host=None):
        try:
            url = await self.client.presigned_get_object(
                self.bucket,
                remote_path,
                change_host=change_host
            )
            print(
                "successfully created download url %s for %s from bucket %s" % (
                    url,
                    remote_path,
                    self.bucket
                )
            )
            return url
        except S3Error as exc:
            print("error occurred.", exc)

File: src/storage.rs
/*
:project: transfery
:author: L-ING
:copyright: (C) 2024 L-ING <hlf01@icloud.com>
:license: MIT, see LICENSE for more details.
*/

use minio::s3::args::{BucketExistsArgs, MakeBucketArgs};
use minio::s3::client::Client;
use minio::s3::creds::StaticProvider;
use minio::s3::http::BaseUrl;

use crate::error;

pub struct Storage {
    client: Client,
    bucket: String,
}

impl Storage {
    pub fn new(
        endpoint: &str,
        username: &str,
        password: &str,
        bucket: &str,
    ) -> Result<Self, error::Error> {
        let base_url = endpoint
            .parse::<BaseUrl>()
            .map_err(|_| error::Error::UrlParseError("Minio endpoint parse failed.".to_string()))?;

        let static_provider = StaticProvider::new(username, password, None);

        let client = Client::new(base_url, Some(Box::new(static_provider)), None, None)
            .map_err(|e| error::Error::MinioClientError(e.to_string()))?;

        Ok(Self {
            client,
            bucket: bucket.to_string(),
        })
    }

    pub async fn init(&self) {
        self.create_buffer_if_not_exists().await;
    }

    async fn create_buffer_if_not_exists(&self) {
        let exists = self
            .client
            .bucket_exists(
                &BucketExistsArgs::new(&self.bucket)
                    .expect("Minio bucket name invalid when checking existence."),
            )
            .await
            .expect("Minio checking bucket existence await failed.");

        if !exists {
            self.client
                .make_bucket(
                    &MakeBucketArgs::new(&self.bucket)
                        .expect("Minio bucket name invalid when making."),
                )
                .await
                .expect("Minio making bucket await failed.");
        }
    }
}
